zip-0320.rst

```diff
@@ -5,7 +5,347 @@
+the text::
+
+  A Unified Address or Unified Viewing Key MUST contain at least one
+  shielded Item (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`).
+  The rationale is that the existing P2SH and P2PKH transparent-only
+  address formats, and the existing P2PKH extended public key format,
+  suffice for representing transparent Items and are already supported
+  by the existing ecosystem.
+
+will be replaced by::
+
+  A Unified Address MUST contain at least one Receiver and any number
+  of Metadata Items. The selection of Receivers is further restricted 
+  such that a Unified Address MUST **either** contain at least one shielded
+  Receiver (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`), OR 
+  MUST contain **only** a Traceable Address Receiver (Typecode
+  :math:`\mathtt{0x04}`).
+
+  A Unified Viewing Key MUST contain at least one shielded Item (Typecodes
+  :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`). 
+
+A Traceable Unified Address is produced from a mainnet Zcash P2PKH address by
+executing the following steps:
+
+1. Decode the address to a byte array using the Base58Check decoding algorithm
+   [#Base58Check]_.
+2. If the length of the resulting byte sequence is not 22 bytes or if its
+   two-byte address prefix is not :math:`\mathtt{[0x1C, 0xB8]}`, return an
+   error. Otherwise, let the **validating key hash** be the remaining 20 bytes
+   of the array after removing the two-byte address prefix.
+3. Construct a new Unified Address using a single Traceable Receiver
+   :math:`\mathtt{0x04}` with the 20-byte **validating_key_hash** as its value.
+   In addition, metadata items such as an Address Expiry Height or Address
+   Expiry Date [#zip-0316-address-expiry]_ MAY be included.
+
+Wallets and other Senders sending to a Traceable Unified Address MUST ensure
+that only transparent UTXOs are spent in the creation of a transaction.
+
+Reference Implementation (Alternative 2)
+----------------------------------------
+
+Javascript using `zcash_address_wasm` [#zcash_address_wasm]_::
+
+    import init, { to_traceable_address, traceable_to_p2pkh, addr_expiry_time } from 'zcash_address_wasm';
+    init().then(() => {
+      var t_address = "t1VmmGiyjVNeCjxDZzg7vZmd99WyzVby9yC";
+      console.log("original p2pkh address: " + t_address);
+    
+      var expiry_time = new Date();
+      expiry_time.setDate(expiry_time.getDate() + 30);
+      var expiry_unix_seconds = BigInt(Math.floor(expiry_time.getTime() / 1000));
+    
+      var traceable_address = to_traceable_address(t_address, expiry_unix_seconds);
+      console.log("traceable address: " + traceable_address);
+    
+      var p2pkh_addr = traceable_to_p2pkh(traceable_address);
+      console.log("decoded p2pkh address: " + p2pkh_addr);
+    
+      var expiry = addr_expiry_time(traceable_address);
+      if (expiry !== null) {
+          console.log("expiry time: " + new Date(Number(expiry) * 1000).toString());
+      }
```

Comment:
Since expiry is a `BigInt` representing a `uint64`, the conversion to `Number` and the multiplication could be imprecise, and the subsequent conversion to `Date` could return an Invalid Date object.

According to the [Mozilla documentation for Date]:
> A Date object can represent a maximum of ±8,640,000,000,000,000 milliseconds, or ±100,000,000 (one hundred million) days, relative to the epoch. This is the range from April 20, 271821 BC to September 13, 275760 AD. Any attempt to represent a time outside this range results in the Date object holding a timestamp value of [NaN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NaN), which is an "Invalid Date".

The potential issue here is not that we directly care about imprecision or invalidity of a far-future expiration time, but that if this code is cut-and-pasted then the Invalid Date object could cause unexpected behaviour in subsequent code. This case is obviously reachable for an adversarially constructed UA.

```suggestion
      // Ignore far-future expiry times not representable as a Date.
      if (expiry !== null && expiry <= 8_640_000_000_000n) {
          console.log("expiry time: " + new Date(Number(expiry) * 1000).toString());
      }
```